src/Psl/Fun/when.php

```diff
@@ -0,0 +1,36 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Fun;
+
+/**
+ * Creates a callback that can run a left or right function based on a condition.
+ *
+ * Example:
+ *
+ *      $greet = Fun\when(
+ *          static fn(string $name): bool => $name === 'Jos',
+ *          static fn(string $name): string => 'Bonjour Jos!',
+ *          static fn(string $name): string => 'Hello ' . $name . '!'
+ *      );
+ *
+ *      $greet('World')
+ *      => Str('Hello World!');
+ *
+ *      $greet('Jos')
+ *      => Str('Bonjour Jos!');
+ *
+ * @template Input
+ * @template Output
+ *
```

Comment:
`Ti` / `To`

usually, we keep templates short and prefix them with `T` ( e.g `Tk` instead of `TKey` / `Tv` instead of `TValue` )